lib/wasix/src/state/func_env.rs

```diff
@@ -73,31 +72,11 @@ impl WasiFunctionEnv {
 
         let mut store = store.unwrap_or_else(|| env.runtime().new_store());
 
-        // Build the context object and import the memory
-        let mut ctx = WasiFunctionEnv::new(&mut store, env);
-        let (mut import_object, init) =
-            import_object_for_all_wasi_versions(&module, &mut store, &ctx.env);
-        if let Some(memory) = memory.clone() {
-            import_object.define("env", "memory", memory);
-        }
-
-        let instance = Instance::new(&mut store, &module, &import_object).map_err(|err| {
-            tracing::warn!("failed to create instance - {}", err);
-            WasiThreadError::InstanceCreateFailed(Box::new(err))
-        })?;
-
-        init(&instance, &store).map_err(|err| {
-            tracing::warn!("failed to init instance - {}", err);
-            WasiThreadError::InitFailed(Arc::new(err))
-        })?;
-
-        // Initialize the WASI environment
-        ctx.initialize_with_memory(&mut store, instance, memory, update_layout)
-            .map_err(|err| {
-                tracing::warn!("failed initialize environment - {}", err);
-                WasiThreadError::ExportError(err)
-            })?;
+        let (_, ctx) =
+            env.instantiate(module, &mut store, memory, update_layout, call_initialize)?;
 
+        // FIXME: shouldn't this happen _before_ instantiating, so the startup code in the instance
+        // has access to the globals?
```

Comment:
I'm not sure this has been used for anything in a while, it's related to deep sleep, I believe.
When restoring, you would NOT want to run the initializer again.

Reply:
I believe it's also used for forking, setjmp, etc. Definitely something we want to keep. But the order really only matters for dl, and dl+asyncify is not a thing yet.